contracts/governance/BaseWeightedMultisig.sol

```diff
@@ -0,0 +1,184 @@
+     * @dev This function checks if the provided signers are sorted and contain no duplicates
+     * @param signers The signers to check
+     * @return True if the signers are sorted and contain no duplicates
+     */
+    function hashForSignersEpoch(uint256 epoch) external view returns (bytes32) {
+        return _baseWeightedStorage().hashForSignersEpoch[epoch];
+    }
+
+    /*
+     * @dev This function checks if the provided signers are sorted and contain no duplicates
+     * @param signers The signers to check
+     * @return True if the signers are sorted and contain no duplicates
+     */
+    function signersEpochForHash(bytes32 hash) external view returns (uint256) {
+        return _baseWeightedStorage().signersEpochForHash[hash];
+    }
+
+    /*************************\
+    |* Integration Functions *|
+    \*************************/
+    /*
+     * @dev This function takes messageHash and proof data and reverts if proof is invalid
+     * @param messageHash The hash of the message that was signed
+     * @param weightedSigners The weighted signers data
+     * @param signatures The signatures data
+     * @return isLatestSigners True if provided signers are the current ones
+     */
+    function _validateProof(
+        bytes32 messageHash,
+        bytes calldata weightedSigners,
+        bytes[] calldata signatures
+    ) internal view returns (bool isLatestSigners) {
+        WeightedMultisigStorage storage $ = _baseWeightedStorage();
+        bytes32 signersHash = keccak256(weightedSigners);
+        WeightedSigners memory signers = abi.decode(weightedSigners, (WeightedSigners));
+        uint256 signersEpoch = $.signersEpochForHash[signersHash];
+        uint256 currentEpoch = $.currentSignersEpoch;
+
+        isLatestSigners = signersEpoch == currentEpoch;
+
+        if (signersEpoch == 0 || currentEpoch - signersEpoch >= OLD_SIGNERS_RETENTION) revert InvalidSigners();
+
+        _validateSignatures(messageHash, signers, signatures);
+    }
+
+    /*
+     * @dev This function rotates the current signers with a new set of signers
+     * @param newWeightedSigners The new weighted signers data
+     */
+    function _rotateSigners(bytes memory newWeightedSigners) internal {
+        WeightedMultisigStorage storage $ = _baseWeightedStorage();
+        WeightedSigners memory newSet = abi.decode(newWeightedSigners, (WeightedSigners));
+
+        // signers must be sorted binary or alphabetically in lower case
+        if (newSet.signers.length == 0 || !_isSortedAscAndContainsNoDuplicate(newSet.signers)) revert InvalidSigners();
+
+        uint256 length = newSet.signers.length;
+        uint256 totalWeight;
+        for (uint256 i; i < length; ++i) {
+            uint256 weight = newSet.signers[i].weight;
+
+            if (weight == 0) revert InvalidWeights();
+
+            totalWeight = totalWeight + weight;
+        }
+        if (newSet.threshold == 0 || totalWeight < newSet.threshold) revert InvalidThreshold();
```

Comment:
This condition `newSet.threshold == 0` could be checked earlier to avoid calling the internal function and performing the loop if the function will revert anyways

Reply:
I totally agree but I think we would prioritize code readability over some gas optimization for a revert scenario. Revert is usually caught by the gas estimation for a TX, so normally no gas is wasted